ledger/src/shred.rs

```diff
@@ -314,16 +314,16 @@ impl Shred {
         }
     }
 
-    pub fn new_from_serialized_shred(mut payload: Vec<u8>) -> Result<Self> {
+    pub fn new_from_serialized_shred(payload: Vec<u8>) -> Result<Self> {
         let mut start = 0;
         let common_header: ShredCommonHeader =
             Self::deserialize_obj(&mut start, SIZE_OF_COMMON_SHRED_HEADER, &payload)?;
 
         let slot = common_header.slot;
-        // Shreds should be padded out to SHRED_PAYLOAD_SIZE
-        // so that erasure generation/recovery works correctly
-        // But only the data_header.size is stored in blockstore.
-        payload.resize(SHRED_PAYLOAD_SIZE, 0);
+        // Data hreds should be padded out to SHRED_PAYLOAD_SIZE so that erasure
+        // generation/recovery works correctly. Check uses >= because a full packet
+        // payload (including the nonce) could be passed in
+        assert!(payload.len() >= SHRED_PAYLOAD_SIZE);
```

Comment:
Looked a little bit more into what we should do here (resize / assertion / etc) after working on something else for a bit. As Carl pointed out, blockstore isn't the only caller of this function. One example of a [non blockstore caller is here](https://github.com/solana-labs/solana/blob/master/core/src/window_service.rs#L250-L255) in `window_service.rs`:
```
                            // shred fetch stage should be sending packets
                            // with sufficiently large buffers. Needed to ensure
                            // call to `new_from_serialized_shred` is safe.
                            assert_eq!(packet.data.len(), PACKET_DATA_SIZE);
                            let serialized_shred = packet.data.to_vec();
                            if let Ok(shred) = Shred::new_from_serialized_shred(serialized_shred) {
```
Quick refresher, `SHRED_PAYLOAD_SIZE == PACKET_DATA_SIZE - SIZE_OF_NONCE`.

In this case, the incoming buffer has the nonce (or at least the space for the nonce) included. If that shred is to subsequently be placed into the blockstore, we do need to drop the "nonce bytes" since we have a check in blockstore where shreds that are too large aren't inserted.

Going to mull on this in the back of my mind and come back to it a bit later - we need a resize / truncate for sure to handle the case I just mentioned, debatable whether we need the assertion.

For reference, here is what we had before:
```
        let expected_data_size = SHRED_PAYLOAD_SIZE;
        // Safe because any payload from the network must have passed through
        // window service,  which implies payload wll be of size
        // PACKET_DATA_SIZE, and `expected_data_size` <= PACKET_DATA_SIZE.
        //
        // On the other hand, if this function is called locally, the payload size should match
        // the `expected_data_size`.
        assert!(payload.len() >= expected_data_size);
        payload.truncate(expected_data_size);
```